particle-iot/device-os
# Replace COPS=2 with CFUN=0 (or CFUN=4) to prevent longer registration times

# Problem

Cell registration takes a long time with certain SIMs in certain locations after using `Cellular.disconnect()` or reset through `particle usb reset`.

### Solution
Gen2 and Gen3: 
- `disconnect()` currently uses `COPS=2` which presumably initiates some kind of rescan on the modem for network, and takes longer time when it's trying to register. Setting the `QCFG` parameters on Quectel modem could also re-trigger scan that could lead to longer reg times.

Gen3:
- `COPS=2` is also done in `initReady()` which again delays the registration. This can be removed as per some testing, `COPS=2` did not look like it is needed. (Please see [this](https://github.com/particle-iot/device-os/pull/2175/files#r473053805) comment) 

Gen2: 
- `COPS=2` is used for setting `CGDCONT` and `URAT` which requires de-registration before being able to set them. This can also be achieved by `AT+CFUN=0` or `AT+CFUN=4` which can be used to de-register before setting these commands, instead of using `COPS=2` with which registration time lengthens. 

### Steps to Test

Check that the connection at Step3 will not take longer than the connection at Step1. 
Do not use `particle usb reset` to reset the module to stat this test, as it triggers as `disconnect()`. 

### Example App

```c
#include "application.h"

SerialLogHandler g_logUsbSerial(LOG_LEVEL_ALL);
SYSTEM_MODE(AUTOMATIC);

void setup(){
    // Step 1: Let a first connection work,
    // and make sure the device is powered on for the first time
    Particle.connect();
    waitUntil(Particle.connected);
    Serial.printlnf("Particle connected");

    // Step 2: Disconnect from cloud and cellular network
    Particle.disconnect();
    waitUntil(Particle.disconnected);
    Cellular.disconnect();
    while(Cellular.ready());

    Serial.printlnf("Cellular disconnected");
    delay(20s);

    // Step 3: 
    Cellular.connect();
    waitUntil(Cellular.ready);
    Serial.printlnf("Cellular connected");
}

void loop() {
}
```

### References

[ch59733](https://app.clubhouse.io/particle/story/59733/quantify-registration-times-in-brn402-between-twilio-and-at-t-sims)

---

### Completeness
